pkg/runtime/wfengine/activity_test.go

```diff
@@ -0,0 +1,57 @@
+/*
+Copyright 2022 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package wfengine_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/google/uuid"
+	"github.com/microsoft/durabletask-go/task"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/dapr/dapr/pkg/actors"
+	"github.com/dapr/dapr/pkg/runtime/wfengine"
+)
+
+// TestDedupeActivityInvocation executes the activity actor invocation path for the same (simulated) workflow
+// twice to verify that the second call fails with a duplication invocation error.
+func TestDedupeActivityInvocation(t *testing.T) {
```

Comment:
Is there any way we can test multiple generations of a workflow happening concurrently? The caching mechanism above (and possibly the storage) seems like it could have race conditions if this were to happen. (Or, if we intend for this to never happen, can we guarantee it doesn't happen?)

Reply:
It's impossible to have two generations of the same workflow or activity to run concurrently since all generations share the same actor ID (and actors already guarantee one-at-a-time execution).